### PR TITLE
executor: fix CTE bug when used with Apply

### LINKED ISSUE #31255: CTE doesn't handle apply properly
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table if exists t1, t2;
    create table t1(c1 int, c2 int);
    insert into t1 values(2, 1);
    insert into t1 values(2, 2);
    create table t2(c1 int, c2 int);
    insert into t2 values(1, 1);
    insert into t2 values(3, 2);
    select * from t1 where c1 > all(with cte1 as (select c1 from t2 where t2.c2 = t1.c2) select c1 from cte1);

### 2. What did you expect to see? (Required)

    mysql>     select * from t1 where c1 > all(with cte1 as (select c1 from t2 where t2.c2 = t1.c2) select c1 from cte1);
    +------+------+
    | c1   | c2   |
    +------+------+
    |    2 |    1 |
    +------+------+
### 3. What did you see instead (Required)

    mysql>     select * from t1 where c1 > all(with cte1 as (select c1 from t2 where t2.c2 = t1.c2) select c1 from cte1);
    +------+------+
    | c1   | c2   |
    +------+------+
    |    2 |    1 |
    |    2 |    2 |
    +------+------+
    2 rows in set (0.00 sec)
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by guo-shaoge**
### What problem does this PR solve?

Issue Number: close #31255

Problem Summary: No handling for Apply. 

### What is changed and how it works?

Add `IsInApply` in `CTEClass` to indicates whether CTE is the inner plan of Apply. If it is, `CTEExec.resTbl` and `CTEExec.iterInTbl` will be reset for each outer row of Apply.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
executor: fix CTE bug when used with Apply
```

### CODE REVIEW

File: executor/explainfor_test.go [hunk 1]
```diff
@@ -1392,7 +1392,7 @@ func (s *testPrepareSerialSuite) TestCTE4PlanCache(c *C) {
 	tk.MustExec("set @a=1, @b=2, @c=3, @d=4, @e=5, @f=0;")
 
 	tk.MustQuery("execute stmt using @f, @a, @f").Check(testkit.Rows("1"))
-	tk.MustQuery("execute stmt using @a, @b, @a").Check(testkit.Rows("1"))
+	tk.MustQuery("execute stmt using @a, @b, @a").Sort().Check(testkit.Rows("1", "2"))
 	tk.MustQuery("select @@last_plan_from_cache;").Check(testkit.Rows("0"))
```
Comment (guo-shaoge) [inline executor/explainfor_test.go:1396]: The correct result is '1 2': 

    drop table if exists t1;
    create table t1(a int);
    insert into t1 values(1);
    insert into t1 values(2);
    SELECT * FROM t1 dt WHERE EXISTS(WITH RECURSIVE qn AS (SELECT a*1 AS b UNION ALL SELECT b+2 FROM qn WHERE b=1) SELECT * FROM qn WHERE b=a);

File: executor/cte_test.go [hunk 2]
```diff
@@ -408,3 +408,20 @@ func TestSpillToDisk(t *testing.T) {
 	}
 	rows.Check(testkit.Rows(resRows...))
 }
+
+func TestCTEWithApply(t *testing.T) {
```
Comment (wjhuang2016) [inline executor/cte_test.go:412]: Please move them to explain_test

Comment (guo-shaoge) [reply executor/cte_test.go]: done

**Review by qw4990 — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**